### PR TITLE
Branch dhruv

### PR DESCRIPTION
**Pull request by dgupta7**
Issues addressed:
1. Capability of have trials spanning different number of DOFs.
2. getShift now runs after set_kT_byName, rather than before.
3. Cost function fixed to have division by the number of mesh points per trial.
4. Scaling parameters coupled muscles are also allowed to change now [fixed variables free_kT, free_lMo and free_lTs]
5. Indexing in eSim variable during MTE corrected.
6. Give user an option to normalize EMG to maxEMG or to (max EMG/max Excitation from dynamic optimization without EMG constraints).
Validations performed already:
1. validated with 4 gait trials. New code vs corrected old code give same results.
2. validated with 2 IPSA trials. New code vs corrected old code give same results.
3. code has been run with a combination of gait and IPSA trials and it ran without errors.

### CODE REVIEW

**Comment by MaartenAfschrift** [comment]
Dhruv, before I start the review can you remove the additional examples. I think we should only include examples of things that are really new (so the example in which you combine different trials should not be in there since we already have this in the repo). This makes it easier to review and we want to keep this repo 'as light as possible' = so only software & minimal data

File: MuscleModel/ReadFromModel/settings_Muscle_analysis_v4.xml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (MaartenAfschrift) [inline MuscleModel/ReadFromModel/settings_Muscle_analysis_v4.xml:1]: why did you add this file ?

File: Optimization/Validation/runValidation.m [hunk 2]
```diff
@@ -0,0 +1,145 @@
+function [Results] = runValidation(Misc,DatStore,Mesh,trial,output,optionssol,Results,NMuscles,lMo_scaling_param_opt,lTs_scaling_param_opt,kT_scaling_param_opt)
```
Comment (MaartenAfschrift) [inline Optimization/Validation/runValidation.m:1]: Is it possible to include this function in the main script again. Now it is impossible to review. I understand that it might be useful to have a function to build the OCP for the three cases. but not a separate function for the three problems. Because now you just copy pasted this into a functions (without saving lines or code because the functions are only used once....).

This could be useful if this function is very generic and can be used in other project. But with the very specific input structure for this project, I really think this is not useful

**Review by MaartenAfschrift — COMMENTED** [review]
some first comments on the PR

File: IO_Functions/GetIndices.m [hunk 3]
```diff
@@ -0,0 +1,127 @@
+function [DatStore,Misc] = GetIndices(DatStore,Misc)
```
Comment (MaartenAfschrift) [inline IO_Functions/GetIndices.m:1]: why is this a new function. Did you rename the function GetIndices_US ?

File: MuscleModel/EMG/GetEMGInfo.m [hunk 4]
```diff
@@ -7,122 +7,136 @@
...
-    iF       = 1;    
-    bool_error  = 0;
-    IndError=zeros(length(Misc.EMGSelection),1);
-    for i=1:length(Misc.EMGSelection)
-        if ~any(strcmp(Misc.EMGSelection{i},DatStore(iF).MuscleNames))
-            disp(['Could not find ' Misc.EMGSelection{i} ' in the model, Update the Misc.EMGSelection structure']);
-            bool_error=1;
-            IndError(i)=1;
+        clear emgFile
+        emgFile = read_motionFile_v40(Misc.EMGfile{iF});
```
Comment (MaartenAfschrift) [inline MuscleModel/EMG/GetEMGInfo.m:28]: why did you change ReadMotFile to read_motionFile_v40 ? Is this one faster (and if so, convince me we need an additional function)

File: MuscleModel/EMG/GetEMGInfo.m [hunk 5]
```diff
@@ -7,122 +7,136 @@
...
+        
+        ct = 0;
+        for i=1:length(Misc.EMGSelection)
+            if any(ismember(DatStore(iF).MuscleNames,Misc.EMGSelection{i}))
+                if any(ismember(EMGheaders{iF},Misc.EMGSelection{i}))
+                    ct = ct +1;
+                    % First column corresponds to index in allMuscleList
+                    % Second column corresponds to index in Misc.EMGheaders
+                    % Third column corresponds to index in MuscleNames of that trial                    
+                    Misc.idx_EMGsel{iF}(ct,1) = find(ismember(Misc.allMuscleList,Misc.EMGSelection{i}));
```
Comment (MaartenAfschrift) [inline MuscleModel/EMG/GetEMGInfo.m:67]: why find find(ismember). And not find(strcmp())

File: MuscleModel/ReadFromModel/getMuscleInfo.m [hunk 6]
```diff
@@ -1,154 +1,162 @@
...
+            Misc.DofNames_muscles{t}{ct}=Misc.DofNames_Input{t}{i};
+            dM_temp(:,i,:)=dM;
+            DOF_inds(ct)=i;
+            ct=ct+1;   
+        end     
     end
-    
-    % Evaluate if one of the muscles spans this DOF (when moment arms > 0.001)
-    dM=dm_Data_temp.data(:,Inds_muscles);    
-    if any(any(abs(dM)>0.001))
```
Comment (MaartenAfschrift) [inline MuscleModel/ReadFromModel/getMuscleInfo.m:46]: we probably want to decrease this treshold. Maybe 0.0001 ?

File: MuscleModel/ReadFromModel/getMuscleInfo.m [hunk 7]
```diff
@@ -1,154 +1,162 @@
...
+    fs=1/mean(diff(t_IK));
+    [B, A] = butter(Misc.f_order_IK, Misc.f_cutoff_IK/(fs/2));
+    [~,DOFS] = size(IK_data.data);
+    for i = 2:DOFS         % filtering time vector not needed (start from 2)
+        IK_data.data(:,i) = filtfilt(B,A,IK_data.data(:,i));
+    end
+
+    %% Filter ID
+    % Get the ID data
+    ID_data=importdata(ID_path);
```
Comment (MaartenAfschrift) [inline MuscleModel/ReadFromModel/getMuscleInfo.m:121]: I think we want to get rid of importdata and want to use ReadMotFile because this is a lot faster !

File: solveMuscleRedundancy.m [hunk 8]
```diff
@@ -26,126 +25,117 @@
...
 % ----------------------------------------------------------------------- %
-tau_act = 0.015;    Misc.tauAct = tau_act * ones(NMuscles, 1);     % activation time constant (activation dynamics)
-tau_deact = 0.06;   Misc.tauDeact = tau_deact * ones(NMuscles,1);  % deactivation time constant (activation dynamics)
+Misc.tau_act = 0.015;    
+Misc.tau_deact = 0.06;   
 Misc.b = 0.1;       % tanh coefficient for smooth activation dynamics
 
-Misc.kT=Misc.kT;
-Misc.shift=Misc.shift;
+% Misc.kT=Misc.kT;
```
Comment (MaartenAfschrift) [inline solveMuscleRedundancy.m:93]: this should be removed and not in comment

File: solveMuscleRedundancy.m [hunk 9]
```diff
@@ -26,126 +25,117 @@
...
+for i=Misc.trials_sel
+    % Static optimization using IPOPT solver (used as an initial guess)
+    DatStore = SolveStaticOptimization_IPOPT_CasADi(DatStore,Misc,i);
 end
 
 %% Input activation and contraction dynamics
 % ----------------------------------------------------------------------- %
-tau_act = 0.015;    Misc.tauAct = tau_act * ones(NMuscles, 1);     % activation time constant (activation dynamics)
-tau_deact = 0.06;   Misc.tauDeact = tau_deact * ones(NMuscles,1);  % deactivation time constant (activation dynamics)
+Misc.tau_act = 0.015;
```
Comment (MaartenAfschrift) [inline solveMuscleRedundancy.m:89]: add this to default settings (that can be changed). Not in the correct place here

File: MuscleModel/ReadFromModel/settings_Muscle_analysis_v4.xml (hunk 1 above)
Comment (dgupta7) [reply MuscleModel/ReadFromModel/settings_Muscle_analysis_v4.xml:1]: I put it there in case version 4 of OpenSim had a different implementation. This file can be removed.

File: IO_Functions/GetIndices.m (hunk 3 above)
Comment (dgupta7) [reply IO_Functions/GetIndices.m:1]: GetIndices_US was already updated to GetIndices_OptMuscleParams. The Getindices I wrote is very similar to GetIndices_OptMuscleParams. Just added the lTs variables, and removed some warnings since they were not needed anymore.

File: MuscleModel/EMG/GetEMGInfo.m (hunk 4 above)
Comment (dgupta7) [reply MuscleModel/EMG/GetEMGInfo.m:28]: No big reason, honestly. The main reason to change it is that the version 4 of OpenSim has a new header variable called inDeg. This new function takes that into account. The inDeg variable is not important in this implementation. Both ReadMotFile and read_motionFile_v40 should work. The structures they output have the same data, but different names for column headers, just need to adjust that.

File: MuscleModel/ReadFromModel/getMuscleInfo.m (hunk 6 above)
Comment (dgupta7) [reply MuscleModel/ReadFromModel/getMuscleInfo.m:46]: done

File: solveMuscleRedundancy.m (hunk 8 above)
Comment (dgupta7) [reply solveMuscleRedundancy.m:93]: done

File: solveMuscleRedundancy.m (hunk 9 above)
Comment (dgupta7) [reply solveMuscleRedundancy.m]: done
